pkg/genyaml/genyaml.go

```diff
@@ -0,0 +1,404 @@
+	JsonTag = "json"
+)
+
+// Comment is an abstract structure for storing mapped types to comments.
+type CommentMap struct {
+	comments map[string]map[string]Comment
+	sync.RWMutex
+}
+
+// NewCommentMap is the constructor for CommentMap accepting a variadic number of paths.
+func NewCommentMap(paths ...string) *CommentMap {
+	cm := &CommentMap{
+		comments: make(map[string]map[string]Comment),
+	}
+
+	for _, path := range paths {
+		cm.AddPath(path)
+	}
+
+	return cm
+}
+
+// Comment is an abstract structure for storing parsed AST comments decorated with contextual information.
+type Comment struct {
+	// Type is the underlying type of the identifier associated with the comment.
+	Type string
+	// IsObj determines if the underlying type is a object type (e.g. struct) or primitive type (e.g. string).
+	IsObj bool
+	// Doc is a comment string parsed from the AST of a node.
+	Doc string
+}
+
+// marshal marshals the object into JSON then converts JSON to YAML and returns the YAML.
+func marshal(o interface{}) ([]byte, error) {
+	j, err := json.Marshal(o)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling into JSON: %v", err)
+	}
+
+	y, err := jsonToYaml(j)
+	if err != nil {
+		return nil, fmt.Errorf("error converting JSON to YAML: %v", err)
+	}
+
+	return y, nil
+}
+
+// jsonToYaml Converts JSON to YAML.
+func jsonToYaml(j []byte) ([]byte, error) {
+	// Convert the JSON to an object.
+	var jsonObj interface{}
+	// We are using yaml.Unmarshal here (instead of json.Unmarshal) because the
+	// Go JSON library doesn't try to pick the right number type (int, float,
+	// etc.) when unmarshalling to interface{}, it just picks float64
+	// universally. go-yaml does go through the effort of picking the right
+	// number type, so we can preserve number type throughout this process.
+	err := yaml3.Unmarshal(j, &jsonObj)
+	if err != nil {
+		return nil, err
+	}
+
+	// marshal this object into YAML.
+	return yaml3.Marshal(jsonObj)
+}
+
+// astFrom takes a path to a Go file and returns the abstract syntax tree (AST) for that file.
+func astFrom(path string) (*doc.Package, error) {
+	fset := token.NewFileSet()
+	m := make(map[string]*ast.File)
+
+	f, err := parser.ParseFile(fset, path, nil, parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse file to AST from path: %s", path)
+	}
+
+	m[path] = f
+	apkg, _ := ast.NewPackage(fset, m, nil, nil)
+
+	astDoc := doc.New(apkg, "", 0)
+	if astDoc == nil {
+		return nil, fmt.Errorf("unable to parse AST documentation from path: %s", path)
+	}
+
+	return astDoc, nil
+}
+
+// fmtRawDoc formats/sanitizes a Go doc string removing TODOs, newlines, whitespace, and various other characters from the resultant string.
+func fmtRawDoc(rawDoc string) string {
+	var buffer bytes.Buffer
+	delPrevChar := func() {
+		if buffer.Len() > 0 {
+			buffer.Truncate(buffer.Len() - 1) // Delete the last " " or "\n".
+		}
+	}
+
+	// Ignore all lines after ---.
+	rawDoc = strings.Split(rawDoc, "---")[0]
```

Comment:
This is behavior taken directly from [kubernetes/kubernetes](github.com/kubernetes/kubernetes) swagger doc generation [script](https://github.com/kubernetes/kubernetes/blob/master/hack/lib/swagger.sh#L44-L46). I have no problem changing if this does not make sense for out use-case.

Reply:
Ah I see what the logic here is.

This is not a standard convention that I know of, but I don't think there's any harm in copying it here and it might be useful sometime.